### ISSUE TITLE
[C] struct members are parsed as typedefs

### ISSUE THREAD

**Issue by jirislaby** [antlr/grammars-v4#1979]
https://github.com/antlr/grammars-v4/blob/4c783b6b0f10927485c4e63007088ddbb3f777c0/c/C.g4#L245-L248
improperly parses:
```
struct X {
        int val1;
};
```
into:
```
(structOrUnionSpecifier 
    (structOrUnion struct) X { 
    (structDeclarationList 
        (structDeclaration 
            (specifierQualifierList 
                (typeSpecifier int) 
                (specifierQualifierList 
                    (typeSpecifier 
                        (typedefName val1)))) ;)) }))
```

Note the `typedefName val1`. val1 is not type specifier, but identifier.

We used GnuC antlr3 grammar in https://github.com/jirislaby/stanse/blob/master/src/cz/muni/stanse/cparser/GNUCa.g with `typedef` handling taken over from https://github.com/antlr/grammars-v3/blob/master/ANSI-C/C.g. I.e. conditional parsing of identifiers. This was eliminated in this v4 grammar, so that the grammar is language agnostic. But it obviously doesn't work here.

**Comment by jirislaby** [comment]
Another example:
```C
(void *)(addr) + 1;
```
results in:
```
(castExpression ( 
    (typeName 
        (specifierQualifierList 
            (typeSpecifier 
                (typeSpecifier void) 
                (pointer *)))) ) 
    (castExpression ( 
        (typeName 
            (specifierQualifierList 
                (typeSpecifier 
                    (typedefName addr)))) ) 
        (castExpression 
            (unaryExpression 
                (unaryOperator +) 
                (castExpression 
                    (unaryExpression 
                        (postfixExpression 
                            (primaryExpression 1))))))))
```
Opposing to:
```C
(void *)addr + 1
```
which results in correct:
```
(additiveExpression 
    (additiveExpression 
        (multiplicativeExpression 
            (castExpression ( 
                (typeName 
                    (specifierQualifierList 
                        (typeSpecifier 
                            (typeSpecifier void) 
                            (pointer *)))) ) 
                (castExpression 
                    (unaryExpression 
                        (postfixExpression 
                            (primaryExpression addr))))))) + 
    (multiplicativeExpression 
        (castExpression 
            (unaryExpression 
                (postfixExpression 
                    (primaryExpression 1))))))
```

**Comment by kaby76** [comment]
[The v4 grammar](https://github.com/antlr/grammars-v4/blob/4c783b6b0f10927485c4e63007088ddbb3f777c0/c/C.g4) comes directly from the [spec](http://www.open-std.org/JTC1/SC22/WG14/www/docs/n1570.pdf). "/** C 2011 grammar built from the C11 Spec */". I checked the parse tree for the first example, and it seems per-spec. I see no alternative through "type-specifier" to derive "int val1" except via "typedef-name". Is there an alternative derivation for "struct X { int val1; };" following the Spec that I don't see?

--Ken

**Comment by kaby76** [comment]
You are absolutely correct on the second example! The order of the rules are incorrect for expressions. These were placed in the order that the spec lists the rules, but that is the wrong order for Antlr! Good catch.

**Comment by jirislaby** [comment]
> I see no alternative through "type-specifier" to derive "int val1" except via "typedef-name". Is there an alternative derivation for "struct X { int val1; };" following the Spec that I don't see?

Sure, like this:
![struct1](https://user-images.githubusercontent.com/530775/109410179-b72b7300-7998-11eb-910b-12393fbce00b.png)

I am not sure typedefs can be handled without semantic predicates.

BTW this is a C grammar derived from antlr4's working on the Linux kernel:
https://github.com/jirislaby/stansec/blob/parser/src/parser/C.g4

**Comment by kaby76** [comment]
You are right. The Spec CFG is ambiguous, but the authors noted "A struct-declaration that does not declare an anonymous structure or anonymous union shall contain a struct-declarator-list". It's not anonymous because the name "X" was used in the context.

The C.g4 grammar will require a semantic predicate (querying whether it's an anonymous struct/union) to disambiguate the choice of `struct-declaration => specifier-qualifier-list struct-declarator-listopt ; => specifier-qualifier-list ; => int val1 ;` vs `struct-declaration => specifier-qualifier-list struct-declarator-listopt ; => specifier-qualifier-list struct-declarator ; => int val1 ;`.

Good catch!

It's likely there are more such problems with C.g4. I see other rules that are suspect, not per-spec. The grammar should be scraped via tool from spec, cleaned up with transforms, and semantic predicates inserted. Trash can do this, but it's in disarray because of other bugs I'm fixing.

--Ken
